Fork

Ek4terin4 opened the pull request:
change num of port

nikarpoff committed:
loging into console and requestUUID added

nikarpoff committed:
Merge pull request #1 from AlexaVita/working-on-log

loging into console and requestUUID added

nikarpoff committed:
logging into file added

.log file added into .gitingore
aop dependency changed in pom.xml
application properties with logging options added
logging into file added

nikarpoff committed:
Merge pull request #2 from AlexaVita/working-on-log

logging into file added

nikarpoff committed:
after success logging added and after throwing logging added

nikarpoff committed:
Merge pull request #3 from AlexaVita/working-on-log

after success method logging added and after throwing logging added

nikarpoff committed:
optimise

nikarpoff committed:
Merge pull request #4 from AlexaVita/working-on-log

optimising

Commit:
change num of port

nikarpoff committed:
Thymeleaf hello world completed

Запрос 'localhost:80/api/hello' возвращает hello_world.html с дефолтным текстом "Hello World!",
 если изменить запрос на '/api/hello?name=MyName', то получим изменённую страницу по шаблону,
которая теперь будет выводить "Hello MyName!"

nikarpoff committed:
Merge pull request #6 from AlexaVita/working-on-th-helloworld

Thymeleaf hello world completed

AlexaVita committed:
added simple authentication with database

AlexaVita committed:
Merge pull request #7 from AlexaVita/basic-security

added simple authentication with database

nikarpoff committed:
logging completed

nikarpoff committed:
Merge pull request #8 from AlexaVita/logging

logging completed

FredPotter committed:
Remove lombok

FredPotter committed:
Remove lombok

FredPotter committed:
Merge remote-tracking branch 'origin/fork' into fork

FredPotter committed:
Remove lombok from logging

AlexaVita committed:
authentication without database

AlexaVita committed:
Merge pull request #9 from AlexaVita/new-basic-security

authentication without database

HappyHDD commented on src/main/java/com/practice/backend/aop/LoggingAspect.java:
```diff
@@ -0,0 +1,38 @@
+/** Логика логирования, вынесенная как сквозная для проекта */
+@Component
+@Aspect
+public class LoggingAspect {
+
+    // Аннотация-advice, которая обеспечивает выполнение метода beforeMainControllerMethod перед каждым вызовом методов,
+    // помеченных в Pointcuts как allMainControllerMethodsWithUUID().
+    // С помощью JoinPoint прокидываем сигнатуру метода и, соответственно, UUID
+    @Before("Pointcuts.allControllersMethodsWithUUID()")
+    public void beforeAllMainControllerMethodsWithUUID(JoinPoint joinPoint) {
+        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
+        Logging.info((UUID) joinPoint.getArgs()[0], methodSignature.getName(), "request start");
+    }
+
+    // После return (а значит и успеха операции) после методов контроллеров
+    @AfterReturning("Pointcuts.allControllersMethodsWithUUID()")
+    public void afterAllMainControllerMethodsWithUUID(JoinPoint joinPoint) {
+        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
+        Logging.info((UUID) joinPoint.getArgs()[0], methodSignature.getName(), "request finished with success");
+    }
+
+    // После ошибок в любом классе контроллере
+    @AfterThrowing(pointcut = "Pointcuts.allControllersMethodsWithUUID()", throwing = "e")
+    public void afterThrowingInAllMainControllerMethodsWithUUID(JoinPoint joinPoint, Exception e) {
+        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
+        Logging.info((UUID) joinPoint.getArgs()[0], methodSignature.getName(), e.getMessage());
+    }
+}
```
Первый пункт UUID мы должны использовать для того, чтобы увидеть некий пользовательский путь (сессию обзывайте, как угодно)  

Пункт второй, помимо UUID вас так же должно интересовать, а какие данные к вам пришли? То бишь хочется увидеть реквест с которым к вам пришли

Пункт третий необходимо не просто вывести, что происходит, а иметь четкое представление, что сделал пользователь, какая последовательность методов отработала

Пункт четвертый нас интересует результат работы, сюда же по хорошему будет добавить, то сколько времени займет отработка того или иного запроса 

Пункт четвертый в идеале логи должны быть разграничены info и debug режим, но это уже пункт со звездой и пока над этим можете не заморачиваться

HappyHDD commented on src/main/java/com/practice/backend/aop/Pointcuts.java:
```diff
@@ -0,0 +1,13 @@
+
+/** Здесь перечислены все точки приложения (методы классов), действия над которыми отслеживаются в Aspects */
+public class Pointcuts {
+
+    /**  Любые методы классов *Controller (любых контроллеров) с любыми модификаторами доступа и возвращаемыми значениями;
+        с любыми параметрами в сигнатуре, но первый параметр всегда типа UUID. */
+    @Pointcut("execution(* com.practice.backend.controller.*Controller.*(java.util.UUID, ..))")
+    public void allControllersMethodsWithUUID() {}
+
+}
```
Очень плохой вариант, таким образом вы делаете контракт, что первым параметром в сигне контроллера, всегда должен быть UUID, что для одного разраба это может быть допустимо, но никак не для команды или продакшен разработки, представь себе лицо человека который будет новеньким у вас в команде, он сделает контроллер в котором он не поставит UUID первым параметром и задаст себе вопрос "А поему у меня не работает логи?" и он потратит время на разбор, почему они не появляются

А если у тебя проект на 200к строк кода? Сколько времени в пустую на этой уйдет? Как сильно у нового сотрудника будет гореть пятая точка?

HappyHDD commented on src/main/java/com/practice/backend/config/RequestUUIDInterceptor.java:
```diff
@@ -0,0 +1,21 @@
+public class RequestUUIDInterceptor implements HandlerInterceptor {
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+        request.setAttribute("requestUUID", UUID.randomUUID());
+        return true;
+    }
```
Не очень одобряю подобны вещи, тут проблемы такие же как и в комментарии выше

Просто подумайте почему данная реализация preHandle она очень очень плохая, потом расскажете

AlexaVita committed:
hikari basic

nikarpoff committed:
Some fixies in logging

Теперь UUID не обязателен как первый параметр request
Логгируются также параметры метода, их названия и входящие аргументы
Логгируется результат запроса и время выполнения

Ek4terin4 committed:
add SQL sript for create DB

Ek4terin4 committed:
add INDEXES and UNIQUE to script BD

AlexaVita committed:
hikari basic

AlexaVita committed:
Merge pull request #11 from AlexaVita/database-test-config

Database test config

nikarpoff committed:
Update MainRestController.java

nikarpoff committed:
Merge pull request #12 from AlexaVita/work-on-logging-again

Some fixies in logging

AlexaVita committed:
example of mybatis

AlexaVita committed:
Merge pull request #14 from AlexaVita/mybatis-test

example of mybatis

Ek4terin4 committed:
Update CreateDB.sql

Fixed some mistakes

nikarpoff committed:
Structure of services and mappers completed

Сделал интерфейс для мапперов с общими запросами
Сделал пример маппера для таблицы Sector (интерфейс и xml)
Написал AbstractService - общий сервис для вызова функций мапперов
Написал пример SectorService
Накинул пару тестовых запросов в контроллер

nikarpoff committed:
Merge pull request #15 from AlexaVita/work-on-mappers-and-services

Structure of services and mappers completed

nikarpoff committed:
Structure of db control

Накинул ещё классов, нужно сделать для каждого маппера запросы и доделать DAO

nikarpoff committed:
Merge pull request #16 from AlexaVita/work-on-mappers-and-services

Structure of db control

bera-m committed:
Update pom.xml

bera-m committed:
Update SectorSettingsMap.java

bera-m committed:
Update Sector.java

bera-m committed:
Update Operation.java

bera-m committed:
Update Fee.java

bera-m committed:
Add files via upload

FredPotter committed:
add Operation.java

FredPotter committed:
Correction of minor errors in DTO, implementation of I...Mapper.xml

FredPotter committed:
Merge pull request #17 from AlexaVita/DTO

Correction of minor errors in DTO (Incorrect variable names, undefined class of operations, duplicate SectorService class), and implementation of I...Mapper.xml

nikarpoff committed:
logging with userUUID

nikarpoff committed:
Merge pull request #18 from AlexaVita/logging

logging with userUUID

nikarpoff committed:
FeeService testing

nikarpoff committed:
Merge pull request #19 from AlexaVita/junit-test

FeeService testing

nikarpoff committed:
PaymentSystem хранится как enum

PaymentSystem теперь как по ТЗ хранится в виде enum в коде, но как строка (имя записи enum) в БД
Добавлено было исключение для ситуации с пустой таблицей
В тестах теперь прокидывается исключение, если мы пытаемся удалить

nikarpoff committed:
Merge pull request #20 from AlexaVita/junit-test

PaymentSystem хранится как enum

nikarpoff committed:
Simple OperationService test

nikarpoff committed:
Merge pull request #21 from AlexaVita/junit-test

Simple OperationService test

BAPanda commented on pom.xml:
```diff
@@ -17,10 +17,39 @@
 <!--			<groupId>org.springframework.boot</groupId>-->
-<!--			<artifactId>spring-boot-starter-security</artifactId>-->
+<!--			<artifactId>spring-boot-starter-data-jpa</artifactId>-->
 <!--		</dependency>-->
```
Если зависимость 100% использоваться не будет - лучше ее сразу выпиливать из pomника, чтобы его не захламлять

BAPanda commented on src/main/java/com/practice/backend/mapper/IMapper.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+
+/** Общий интерфейс для всех Mappers */
+public interface IMapper<EntityType> {
```
здесь нужно ограничить EntityType. сейчас получается, что подсунуть сюда можно любой тип, но по факту маппер должен работать только с классами, у которых определен интерфейс IGuapEntity. пример возможного косяка: в методе insert делаем проверку, что вставка успешна проверкой entity.getId() != null, создаем маппер от класса, у которого нет getId(). на компиляции падаем с ошибкой

BAPanda commented on src/main/java/com/practice/backend/model/IGuapEntity.java:
```diff
@@ -0,0 +1,7 @@
+public interface IGuapEntity {
+    Long getId();
+
+    Long getSectorId();
```
я бы ее отсюда выпилила. сейчас у вас все сущности подразумевают наличие поля sectorId, но в дальнейшем могут появится такие, где сектора нет, и для них этот интерфейс будет неприменим, хотя они все еще будут, по сути дела, GuapEntity. такие маркерные интерфейсы нужно разделять и по возможности (если не требуется иного), оставлять в них только одну сигнатурную функцию

BAPanda commented on src/main/java/com/practice/backend/model/Sector.java:
```diff
@@ -0,0 +1,111 @@
+    private String signCode;
+    private Boolean checkIp;
+    @Pattern(regexp = "[\\p{Print}]{0,255}")
+    private String allowedIps;
```
если уж использовать валидацию с паттернами, то тогда довести дело до ума: сделать паттерн который не только проверит размер строки айпишников, но и валидность самих айпишников

BAPanda commented on src/main/java/com/practice/backend/model/SectorSettingsMap.java:
```diff
@@ -0,0 +1,121 @@
+        this.value = value;
+    }
+
+    public Boolean getPayment() {
+        return payment;
+    }
+
+    public void setPayment(Boolean payment) {
+        this.payment = payment;
+    }
+
+    public String getBinList() {
+        return binList;
+    }
+
+    public void setBinList(String binList) {
+        this.binList = binList;
+    }
+
+    public Boolean getGuapPay() {
+        return guapPay;
+    }
+
+    public void setGuapPay(Boolean guapPay) {
+        this.guapPay = guapPay;
+    }
```
по постановке это не поля таблицы. это имена возможных настроек. вам нужно сделать enum, в котором будут три настройка: binList, payment и guapPay. имена объектов этого enum буду записываться в поле name сущности sectorSettingMap

nikarpoff replied:
То есть эти поля, геттеры и сеттеры удалить? И потом в коде их просто доставать из имени? И можете, пожалуйста, объяснить, что вообще значат эти binList, payment, guapPay))? Они могут быть в какой-нибудь комбинации в имени, или там всегда один (или ни один) из них?

BAPanda commented on src/main/java/com/practice/backend/model/Sector.java:
```diff
@@ -0,0 +1,111 @@
+                ", checkIp=" + checkIp +
+                ", allowedIps='" + allowedIps + '\'' +
+                '}';
+    }
```
signCode выводить в toString ни в коем случае нельзя! это "соль" для подписи запроса. нельзя знать заранее, кто, когда и где решит вызвать этот метод. при этом вызове в логи/сообщение/... попадет signCode и если кто-то сможет получить доступ к этим условным логам, будет утечка конфиденциальной информации со всем вытекающими

BAPanda commented on src/main/java/com/practice/backend/model/operation/Operation.java:
```diff
@@ -0,0 +1,164 @@
+    @NotNull
+    private Long id;
+    @NotNull
+    private Long sector_id;
```
имя поля должно быть в camel case

BAPanda commented on src/main/resources/application.yml:
```diff
@@ -1,4 +1,22 @@
 server:
-  port: 8080
+  port: 80
```
не очень хорошее решение. 80 порт по умолчанию может быть занят только под root'ом. соответственно, для поднятия приложения нужно будет шаманить с правами или с переадресацией. использование не-рутовых портов - правило хорошего тона. кроме того они с меньшей вероятностью окажутся заняты и никаких дополнительных танцев с бубнами не потребуется

BAPanda commented on src/main/java/com/practice/backend/mapper/IMapper.java:
```diff
@@ -0,0 +1,19 @@
+
+    void delete(Long id);
+
+    void update(Long updatedId, EntityType entity);
```
не очень понятно, зачем отдельно выносить updatedId. в entity у вас уже должен быть id, используйте его

BAPanda commented on src/main/resources/com/practice/backend/mapper/IFeeMapper.xml:
```diff
@@ -0,0 +1,32 @@
+            percent = #{entity.percent},
+            fix = #{entity.fix},
+            notless = #{entity.notLess}
+        WHERE id = #{updatedId};
```
при передаче в качестве параметра составного типа (то бишь объекта), можно просто указывать имена полей, без обращения через entity. + учесть замечание про updatedId
относится ко всем мапперам

BAPanda commented on src/main/resources/com/practice/backend/mapper/IOperationMapper.xml:
```diff
@@ -0,0 +1,35 @@
+            fee = #{entity.fee},
+            description = #{entity.description},
+            email = #{entity.email},
+            state = #{entity.state},
+            type = #{entity.type}
```
state и type - перечисления. батис может не понять, как именно вы хотите преобразовать их в значение для БД. здесь и для всех enum'ов нужно указать jdbcType, javaType и typeHandler. у батиса есть дефолтный EnumTypeHandler, который пытается искать по имени объекта енама. но лучше перестраховаться и явно указать: хотите ли вы использовать имена или ординалы. 
относится ко всем местам, где в запросах используются енамы

BAPanda commented on src/test/java/com/practice/backend/service/FeeServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+        }
+
+        // Генерируем обновлённую на основе имеющейся (рандомим fix и notLess)
+        Fee updated = new Fee(initial.getId(), initial.getSectorId(), initial.getPaymentSystem(),
+                initial.getPercent(), Integer.toString(random.nextInt(1000)), Integer.toString(random.nextInt(1000)));
+
+        // Используем сервис и обновляем запись
+        feeServiceTest.update(updated.getId(), updated);
```
как раз про updatedId. intial.getId() равен update.getId(), вытаскивать его отдельно не нужно, он уже есть в объекте updated

BAPanda commented on src/main/java/com/practice/backend/model/Sector.java:
```diff
@@ -0,0 +1,111 @@
+                Objects.equals(allowedIps, sectorDto.allowedIps);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name, active, signCode, checkIp, allowedIps);
+    }
```
не обязательно. если вы нигде не собираетесь сравнивать объекты через хэш и нет необходимость гарантировать правильность этого сравнения, переопределять hashCode не нужно
относится ко всем сущностям

nikarpoff committed:
pom optimised, .gitignore optimised

nikarpoff committed:
code review fixes

Джейнерик EntityType ограничен реализациями интерфейса IGuapEntity
Для enum теперь предназначена отдельная папка
IGuapEntity больше не содержит getSectorId()
sector_id в Operation переведён в camel case

nikarpoff committed:
hashCode removed from entities

В сущностях отныне не переопределён hashCode

nikarpoff committed:
signCode removed from toString() method

nikarpoff committed:
update operation optimised

Ek4terin4 committed:
change port

Ek4terin4 committed:
rename enums

Ek4terin4 committed:
create new enum and delete extra fields from SectorSettingMap

bera-m committed:
Update Sector.java

Validity of an IP address

Ek4terin4 committed:
remove useless fields from DB table and Mapper.xml

Ek4terin4 committed:
Merge remote-tracking branch 'origin/fork' into fork

AlexaVita committed:
tomcat max requests

bera-m committed:
fix regular errors

nikarpoff committed:
Function checkSignature added

Создал Util
Добавил и реализовал статический метод checkSignature
Добавил общее исключение PaymentException

nikarpoff committed:
Merge pull request #22 from AlexaVita/working-on-util

Function checkSignature added

AlexaVita committed:
added checkPan

nikarpoff committed:
Update PaymentUtil.java

Добавил проверку карты на отсутствие символов
Добавил метод checkIp
Добавил метод getPanMask

Ek4terin4 committed:
add field pan_mask in Operation

AlexaVita committed:
changed PaymentUtil

Ek4terin4 committed:
add field pan_mask in Operation (DB and Mapper)

AlexaVita committed:
Payment begin

nikarpoff committed:
getEncodedSignature and buildSignature added

nikarpoff committed:
Merge pull request #23 from AlexaVita/util-update

getEncodedSignature and buildSignature added

AlexaVita committed:
Payment begin

AlexaVita committed:
Merge branch 'fork' of https://github.com/AlexaVita/backend-practice into fork

AlexaVita committed:
Payment begin

AlexaVita committed:
Payment begin

nikarpoff committed:
some additional things

Добавил простейший фасад над PaymentUtil для сокращения кода для проверок
Сделал pojo класс со всеми необходимыми параметрами, пока не используется

nikarpoff committed:
Merge pull request #24 from AlexaVita/working-on-payment

some additional things

Ek4terin4 committed:
A separate method for the Luhn algorithm. Correction for checking boolean.

nikarpoff committed:
fixes

nikarpoff committed:
notify and error added

Теперь при ошибках и успехе операции выдаётся страница с фронта

FredPotter committed:
move mapper to dao.mapper

AlexaVita committed:
kafka

nikarpoff committed:
Random kafka added

Приложение теперь рандомит статус
Если статус операции TIMEOUT, то оно спит 20 секунд, из-за чего мы обрабатываем ошибку о задержке (ну или можно повысить время ожидания кафки)
Затем считывается номер карты, проверяется, в бд записывается результат операции.
Происходит редирект на нотифай с полученным ответом
Нотифай теперь достаёт из базы данных операцию и подставляет её поля как результат в шаблон

nikarpoff committed:
Merge pull request #26 from AlexaVita/working-on-kafka-random

Random kafka added

nikarpoff committed:
Оптимизировал Notify, теперь аргументы прокидываются через pojo

nikarpoff committed:
Merge pull request #27 from AlexaVita/optimising-notify

Оптимизировал Notify, теперь аргументы прокидываются через pojo

nikarpoff committed:
Checking in notify added

Теперь и в notify происходит проверка валидности сигнатуры, IP и активности сектора

nikarpoff committed:
Admin controller work starts

Начал делать админку, два запроса готовы
Теперь в ISectorSettingsMapMapper и в FeeMapper можно искать значения по sectorId
В operation была добавлена paymentSystem
Убрано исключение об пустой таблице из getAll в dao, возвращается просто пустой список
Добавлена настройка для сектора - show_email, для фронта
Добавлен метод расчёта комиссии в отдельном классе OperationUtil

nikarpoff committed:
admin requests

nikarpoff committed:
changed structure of admin requests

Добавил пару запросов, отредактировал старые, добавил требующуюся настройку сектора - storeLogo
